web/api/v1/api.go

```diff
@@ -481,6 +492,72 @@ func (api *API) queryRange(r *http.Request) (result apiFuncResult) {
 	}, nil, res.Warnings, qry.Close}
 }
 
+func (api *API) queryExemplars(r *http.Request) apiFuncResult {
+	start, err := parseTimeParam(r, "start", minTime)
+	if err != nil {
+		return invalidParamError(err, "start")
+	}
+	end, err := parseTimeParam(r, "end", minTime)
+	if err != nil {
+		return invalidParamError(err, "end")
+	}
+	if end.Before(start) {
+		err := errors.New("end timestamp must not be before start time")
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+
+	ctx := r.Context()
+	expr, err := parser.ParseExpr(r.FormValue("query"))
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+
+	mint, maxt := api.QueryEngine.FindMinMaxTime(&parser.EvalStmt{Expr: expr, Start: start, End: end, Interval: 15 * time.Second})
+	q, err := api.Queryable.Querier(ctx, mint, maxt)
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	defer q.Close()
+
+	selectors, err := parser.ExtractSelectors(expr)
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	if len(selectors) < 1 {
+		return apiFuncResult{nil, nil, nil, nil}
+	}
+
+	var retExemplars []exemplarData
+	var series []storage.SeriesSet
+
+	for _, selectorArr := range selectors {
+		series = append(series, q.Select(false, nil, selectorArr...))
+	}
+
+	for _, s := range series {
```

Comment:
nit: why do you build a list of SeriesSets just to iterate over them again?  Can these two loops be merged?